File: tests/oracle/printc_pending_brace_emit_1204.cc
/*
 * PRINTC-STRUCTEMIT-MAIN-IVAR4-DUP-0001: locked Ghidra 12.0.4 PrintC::
 * emitBlockIf PendingBrace oracle (printc.cc:2878-2948 + prettyprint.hh:
 * 102/443-457/1129-1137):
 *   if (isSet(pending_brace)) emit->setPendingPrint(&pendingBrace);   2884
 *   ...condBlock->emit(this)...                                       2897
 *   if (emit->hasPendingPrint(&pendingBrace)) {                       2900
 *     emit->cancelPendingPrint(); emit->spaces(1);                    2901-2902
 *   } else emit->tagLine();                                           2905
 *   ...                                                               2946
 *   if (pendingBrace.getIndentId() >= 0)
 *     emit->closeBraceIndent(CLOSE_CURLY, pendingBrace.getIndentId());2947-2948
 * The PendPrint slot is BASE-CLASS Emit state; only EmitPrettyPrint::
 * tagLine fires it (prettyprint.cc:920/930) — EmitNoMarkup::tagLine
 * (prettyprint.hh:557) never does.
 *
 * Mirrors tests/oracle/printc_pending_brace_emit_1204.rs case for case.
 * Structure built like the whiledo fixture (BlockCopy mirrors via
 * newBlockCopy/copymap/replaceUsingMap as buildCopy block.cc:1925-1938,
 * composites via identifyInternal like ruleBlockProperIf): a 3-component
 * parent properif(condA CBRANCH 3, thenA RET 20, else = child properif)
 * where the child's condition block varies per case.
 *
 * Cases:
 *   1. elseif_stmtcond_pretty — child condition = [RET 21, CBRANCH 9],
 *      EmitPrettyPrint: the condition statement's tagLine FIRES the
 *      pending brace -> `else {` + `return 21;` + newline `if (9) {` +
 *      deferred close (cc:2946-2948). The main-region golden shape.
 *   2. elseif_emptycond_pretty — child condition = [CBRANCH 9] only,
 *      EmitPrettyPrint: nothing fires the brace -> cc:2900-2902 cancel +
 *      spaces(1) -> merged `else if (9) {`, no deferred close.
 *   3. elseif_stmtcond_nomarkup — same graph as case 1, EmitNoMarkup:
 *      tagLine never fires the brace (hh:557) -> condition statements
 *      print inline after `else`, then cc:2902 spaces(1) merges the
 *      `if` onto the statement's line; indentId stays -1 so cc:2946
 *      closes nothing.
 */

#include <bits/stdc++.h>

// Test-only access (same scheme as printc_whiledo_body_emit_1204.cc).
#define private public
#define class struct
#include "architecture.hh"
#include "block.hh"
#include "blockaction.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "printc.hh"
#include "prettyprint.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef class
#undef private

using namespace ghidra;

class FixturePrintC final : public PrintC {
public:
  explicit FixturePrintC(bool pretty)
    : PrintC(nullptr, "printc-pending-brace-emit-1204") {
    delete emit;
    emit = pretty ? static_cast<Emit *>(new EmitPrettyPrint())
                  : static_cast<Emit *>(new EmitNoMarkup());
  }

  void render(const BlockIf *bl) {
    // docFunction (printc.cc:2651/2664) opens the emitter's function group
    // before any block emission — EmitPrettyPrint's indent stack needs that
    // root group (TokenBreak/BeginIndent read indentstack.back()). The group
    // itself prints no bytes (PrintClass::Begin), so the observation stays
    // the bare if-emission stream.
    int4 id = emit->beginFunction((const Funcdata *)0);
    emitBlockIf(bl);
    emit->endFunction(id);
    emit->flush();
  }
};

static void append(BlockBasic &block, PcodeOp *op)
{
  block.insert(block.endOp(), op);
}

struct PendingBraceBuild {
  PcodeOpBank bank;
  ConstantSpace constant_space;
  TypeBase int_type;
  TypeOpCbranch cbranch_type;
  TypeOpReturn return_type;

  BlockBasic conda{nullptr};		// parent condition basic
  BlockBasic thena{nullptr};		// parent then-branch
  BlockBasic condb{nullptr};		// child condition basic
  BlockBasic thenb{nullptr};		// child then-branch

  BlockGraph structure;
  BlockIf *parent;

  explicit PendingBraceBuild(bool stmt_in_cond)
    : constant_space(nullptr, nullptr), int_type(4, TYPE_INT),
      cbranch_type(nullptr), return_type(nullptr),
      parent((BlockIf *)0)
  {
    append(conda, make_cbranch(3));
    append(thena, make_return(20));
    if (stmt_in_cond)
      append(condb, make_return(21));
    append(condb, make_cbranch(9));
    append(thenb, make_return(40));
  }

  Varnode *bool_const(uintb value)
  {
    Varnode *vn = new Varnode(1, Address(&constant_space, value), &int_type);
    new HighVariable(vn);
    return vn;
  }

  PcodeOp *make_cbranch(uintb value)
  {
    PcodeOp *op = bank.create(2, Address());
    bank.changeOpcode(op, &cbranch_type);
    op->setInput(bool_const(0x1000), 0);
    op->setInput(bool_const(value), 1);
    return op;
  }

  PcodeOp *make_return(uintb value)
  {
    Varnode *val = new Varnode(4, Address(&constant_space, value), &int_type);
    new HighVariable(val);
    PcodeOp *op = bank.create(2, Address());
    bank.changeOpcode(op, &return_type);
    op->setInput(bool_const(0x1000), 0);
    op->setInput(val, 1);
    return op;
  }

  void finish(void)
  {
    std::vector<FlowBlock *> originals;
    originals.push_back(&conda);
    originals.push_back(&thena);
    originals.push_back(&condb);
    originals.push_back(&thenb);

    std::vector<BlockCopy *> copies;
    for (int4 i = 0; i < (int4)originals.size(); ++i) {
      BlockCopy *c = structure.newBlockCopy(originals[i]);
      originals[i]->copymap = c;
      copies.push_back(c);
    }
    for (int4 i = 0; i < (int4)copies.size(); ++i)
      copies[i]->replaceUsingMap();

    BlockIf *child = new BlockIf();
    std::vector<FlowBlock *> childnodes;
    childnodes.push_back(copies[2]);		// child condition
    childnodes.push_back(copies[3]);		// child then
    structure.identifyInternal(child, childnodes);
    structure.addBlock(child);

    parent = new BlockIf();
    std::vector<FlowBlock *> parentnodes;
    parentnodes.push_back(copies[0]);		// parent condition
    parentnodes.push_back(copies[1]);		// parent then
    parentnodes.push_back(child);		// parent else = child properif
    structure.identifyInternal(parent, parentnodes);
    structure.addBlock(parent);
  }

  std::string render(bool pretty)
  {
    std::ostringstream output;
    FixturePrintC printer(pretty);
    printer.setOutputStream(&output);
    printer.render(parent);
    return output.str();
  }
};

static std::string to_hex(const std::string &value)
{
  std::ostringstream stream;
  stream << std::hex << std::setfill('0');
  for (const unsigned char byte : value)
    stream << std::setw(2) << static_cast<unsigned int>(byte);
  return stream.str();
}

int main(int argc, char **argv)
{
  const bool raw = argc == 2 && std::string(argv[1]) == "--raw";
  {
    // 1. elseif_stmtcond_pretty
    PendingBraceBuild s(true);
    s.finish();
    std::cout << "elseif_stmtcond_pretty=" << (raw ? to_hex(s.render(true)) : s.render(true)) << '\n';
  }
  {
    // 2. elseif_emptycond_pretty
    PendingBraceBuild s(false);
    s.finish();
    std::cout << "elseif_emptycond_pretty=" << (raw ? to_hex(s.render(true)) : s.render(true)) << '\n';
  }
  {
    // 3. elseif_stmtcond_nomarkup
    PendingBraceBuild s(true);
    s.finish();
    std::cout << "elseif_stmtcond_nomarkup=" << (raw ? to_hex(s.render(false)) : s.render(false)) << '\n';
  }
  return 0;
}

File: tests/oracle/printc_pending_brace_emit_1204.rs
// PRINTC-STRUCTEMIT-MAIN-IVAR4-DUP-0001: Rugra side of the locked Ghidra
// 12.0.4 PrintC::emitBlockIf PendingBrace bilateral fixture (printc.cc:
// 2878-2948 + prettyprint.hh:102/443-457/1129-1137). Mirrors
// printc_pending_brace_emit_1204.cc case for case: hand-built 3-component
// parent BlockIf (condition CBRANCH const, then RETURN, else = child
// BlockIf) driven through PrintC::emit_block_graph with either an
// EmitPrettyPrint (cases 1-2, the production emitter per printlanguage.cc:
// 69) or an EmitNoMarkup (case 3), locking:
//   1. elseif_stmtcond_pretty — child condition [RET 21, CBRANCH 9]: the
//      pending brace FIRES at the condition statement's tagLine ->
//      `else {` + `return 21;` + newline `if (9) {` + deferred close.
//   2. elseif_emptycond_pretty — child condition [CBRANCH 9] only:
//      nothing fires the brace -> cancel + spaces(1) -> `else if (9) {`.
//   3. elseif_stmtcond_nomarkup — case-1 graph with EmitNoMarkup: the
//      slot stays pending (hh:557 never fires) -> statements inline after
//      `else`, spaces(1) merges the `if` onto the statement's line, and
//      no deferred close.

use rugra::address::{Address, SeqNum};
use rugra::block::{BlockBasic, BlockGraph, BlockIf};
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::prettyprint::{Emit, EmitNoMarkup, EmitPrettyPrint};
use rugra::printc::PrintC;
use rugra::space::AddressSpace;
use rugra::varnode::Varnode;
use std::sync::{Arc, RwLock};

type BlockArc = Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>;

fn const_varnode(size: usize, value: u64) -> Arc<RwLock<Varnode>> {
    Arc::new(RwLock::new(Varnode::new_with_space(size, AddressSpace::Const, value)))
}

fn make_op(index: u32, opcode: OpCode) -> PcodeOpRef {
    let mut op = PcodeOp::new(SeqNum::new(Address::new(index as u64), index), opcode);
    op.set_opcode_flags(opcode);
    PcodeOpRef(Arc::new(RwLock::new(op)))
}

fn make_cbranch(op_index: u32, value: u64) -> PcodeOpRef {
    let op = make_op(op_index, OpCode::CPUI_CBRANCH);
    {
        let mut o = op.0.write().unwrap();
        o.inrefs.push(const_varnode(1, 0x1000));
        o.inrefs.push(const_varnode(1, value));
    }
    op
}

fn make_return(op_index: u32, value: u64) -> PcodeOpRef {
    let op = make_op(op_index, OpCode::CPUI_RETURN);
    {
        let mut o = op.0.write().unwrap();
        o.inrefs.push(const_varnode(1, 0x1000));
        o.inrefs.push(const_varnode(4, value));
    }
    op
}

fn basic_block(index: i32, ops: Vec<PcodeOpRef>) -> BlockArc {
    let mut block = BlockBasic::new(index, Address::new(0x1000 + index as u64 * 0x10));
    block.ops = ops;
    Arc::new(RwLock::new(block))
}

// Parent properif(condA CBRANCH 3, thenA RET 20, else = child properif).
// stmt_in_cond: child condition = [RET 21, CBRANCH 9] vs [CBRANCH 9].
fn build_pending_brace(stmt_in_cond: bool) -> BlockGraph {
    let conda = basic_block(0, vec![make_cbranch(1, 3)]);
    let thena = basic_block(1, vec![make_return(2, 20)]);
    let condb = if stmt_in_cond {
        basic_block(2, vec![make_return(3, 21), make_cbranch(4, 9)])
    } else {
        basic_block(2, vec![make_cbranch(4, 9)])
    };
    let thenb = basic_block(3, vec![make_return(5, 40)]);

    let child: BlockArc = Arc::new(RwLock::new(BlockIf {
        index: 4,
        condition: condb,
        if_body: thenb,
        else_body: None,
        goto_target: None,
        goto_type: 0,
        incoming: Vec::new(),
        outgoing: Vec::new(),
        parent: None,
        flags: 0,
    }));

    let parent: BlockArc = Arc::new(RwLock::new(BlockIf {
        index: 5,
        condition: conda,
        if_body: thena,
        else_body: Some(child),
        goto_target: None,
        goto_type: 0,
        incoming: Vec::new(),
        outgoing: Vec::new(),
        parent: None,
        flags: 0,
    }));

    let mut graph = BlockGraph::new();
    graph.add_block(parent);
    graph
}

fn render_pretty(graph: &BlockGraph) -> String {
    // docFunction (printc.cc:2651/2664) opens the emitter's function group
    // before any block emission — EmitPrettyPrint's indent stack needs that
    // root group (token commit reads indentstack.last()). The group prints
    // no bytes, so the observation is the bare if-emission stream.
    let mut emitter = EmitPrettyPrint::new();
    emitter.begin_function();
    let mut printer = PrintC::new(Box::new(emitter));
    printer.emit_block_graph(graph);
    let mut emit = printer
        .take_emit()
        .into_any()
        .downcast::<EmitPrettyPrint>()
        .expect("PrintC fixture must retain EmitPrettyPrint");
    emit.end_function();
    emit.get_output()
}

fn render_nomarkup(graph: &BlockGraph) -> String {
    let mut emitter = EmitNoMarkup::new();
    emitter.begin_function();
    let mut printer = PrintC::new(Box::new(emitter));
    printer.emit_block_graph(graph);
    let mut emit = printer
        .take_emit()
        .into_any()
        .downcast::<EmitNoMarkup>()
        .expect("PrintC fixture must retain EmitNoMarkup");
    emit.end_function();
    emit.flush();
    emit.debug_get_output_ref().to_owned()
}

fn to_hex(value: &str) -> String {
    value
        .as_bytes()
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect()
}

fn emit_case(name: &str, raw_text: &str, hex: bool) {
    if hex {
        println!("{name}={}", to_hex(raw_text));
    } else {
        println!("{name}={raw_text}");
    }
}

fn main() {
    let raw = std::env::args().nth(1).as_deref() == Some("--raw");

    // 1. elseif_stmtcond_pretty
    {
        let graph = build_pending_brace(true);
        emit_case("elseif_stmtcond_pretty", &render_pretty(&graph), raw);
    }

    // 2. elseif_emptycond_pretty
    {
        let graph = build_pending_brace(false);
        emit_case("elseif_emptycond_pretty", &render_pretty(&graph), raw);
    }

    // 3. elseif_stmtcond_nomarkup
    {
        let graph = build_pending_brace(true);
        emit_case("elseif_stmtcond_nomarkup", &render_nomarkup(&graph), raw);
    }
}
